### DISCUSSION TITLE
how to get the PCID on X86

### DISCUSSION THREAD

**Discussion by ai006** [gem5/gem5#1848] [Q&A]
I am trying to retrieve the PCIDs of different processes during FS mode. According to the  Intel 64 and IA-32 Architectures Software Developer’s Manual, PCIDs are described as follows:  

> **PCID: Process-context identifiers.**  
> If `CPUID.01H:ECX.PCID` [bit 17] = 1, the `CR4.PCIDE` flag can be set to 1, enabling process-context identifiers.  

Essentially, if the `CPUID` instruction returns a value with bit 17 asserted, it indicates that PCIDs are supported, and the `CR4.PCIDE` flag can be set to enable this feature.  

In the file `src/arch/x86/fs_workload.cc`, I attempted to enable PCIDs by setting `CR4.PCIDE = 1` within the `initState()` function. However, when I read the PCID, it still returns zero.  

I suspect the issue might be related to CPU configuration. I am currently examining the `X86IntelMPProcessor` in `X86Board.py` to determine how to properly enable PCIDs, but so far, I have not had any success.  

Could someone help me with retrieving PCIDs?

**Comment by ZDeHn** [comment]
I'm currently working on TLB development and encountered the same issue as you. Has your problem been resolved yet? Or does GEM5 not support PCID at all?

**Reply by ai006** [reply]
I think it should support it, but I wasn't able to get it to work. There are alternative solutions, if your work can be done on the ARM ISA, you should use it. ARM has its own version of PCID called ASID, and it works. However, if you need to stay on X86, you can use the root page table address as a form of PID. 

see src/arch/x86/pagetable_walker.cc inside in setupWalk function

**Comment by ZDeHn** [comment]
Thank you very much for your response. If I succeed, I will share my specific solution with you. Could you please help me with another question? The purpose of PCID is to reduce TLB flushes during context switches, but PCID does not appear to work without modifying GEM5. Therefore, by default, does GEM5 flush the TLB during context switches? What is the timing and scope of the flush?
